### PR TITLE
feat: Add hash map

### PR DESCRIPTION
**Pull request by rureverek**
- Unique Table implementation: Migrate from vector to unordered_map.

### CODE REVIEW

**Comment by rureverek** [comment]
This one is now working. We can insert values using `emplace()` method and look for using find() or Table[key].
GPT suggests to use `find()` it that way:
```C++
    auto it = myHashMap.find(keyToFind);

    if (it != myHashMap.end()) {
        std::cout << "Value: << it->second << std::endl;
    } else {
        std::cout << "Key not found." << std::endl;
    }
```
Note that using operator[] has the side effect of inserting a new element into the map with the specified key if the key does not exist. Therefore, when using operator[], you should be careful to ensure that the key exists in the map or use find to check if the key is present before accessing it with operator[].

Next, we need to change the isVariable, isConstant and createVar functions. Now, you can commit to this branch to fix those functions: @emanuelacm @Adhaywe

**Comment by rureverek** [comment]
Also note for `emplace()`
The insertion only takes place if no other element in the container has a key equivalent to the one being emplaced (keys in a [map](https://cplusplus.com/map) container are unique).

**Comment by rureverek** [comment]
@emanuelacm I think we can also move hash_map struct inside Manager class. We are not using templates and I think it should run now.

**Review by Adhaywe — APPROVED** [review]

File: src/Manager.cpp [hunk 1]
```diff
@@ -70,13 +70,23 @@ namespace ClassProject {
...
+   /**
+     * @brief isVariable
+     *
+     * checks whether the item corresponding to a given ID is a variable or not.
+     * @param x: ID of the node.
+     * @return returns true if the ID represents a variable.
+     */
+    bool Manager::isVariable( BDD_ID x) {
+        Unique_Table_Key key{};
+        auto it = Table.find(key);
```
Comment (rureverek) [inline src/Manager.cpp:82]: To find given id:
```C++
auto it = std::find_if(std::begin(Table), std::end(Table),
                         [](auto&& p) { return p->second.id == x; });
```
or
```C++
for (HandleMap::const_iterator it = Table.begin(); it != Table.end(); ++it) {
  if (it->second.it == x) ...
} 
```
See: https://stackoverflow.com/questions/32780046/finding-an-element-in-map-by-its-value

I think yours won't work for us.

**Comment by Adhaywe** [comment]
I will check it.

**Review by rureverek — COMMENTED** [review]
I think you accidentaly add some Cmake files... I'll fix it and this variable function too

**Comment by Adhaywe** [comment]
About the cmake files, don't know how they are pushed but it can be deleted
I guess.I have tested the function with different test cases, it is
returning the correct results. However, if there is still a mistake, I am
happy if you could review it and let me know.

**Comment by rureverek** [comment]
> About the cmake files, don't know how they are pushed but it can be deleted I guess.I have tested the function with different test cases, it is returning the correct results. However, if there is still a mistake, I am happy if you could review it and let me know.

Sorry, but i had to revert commit to delete those cmake files and I just wrote my own, hope its okay.
I think your funcion was okay, but I wasn't sure about this find_if function. If you understand it then we can use your version.

**Comment by Adhaywe** [comment]
No problem at all. I have seen the changes and yours look clean. Thanks for
pointing that out.

**Pull request merged**